### ISSUE TITLE
Next network is tested before current one fails/promotes

### ISSUE THREAD

**Issue by WhiteHalmos** [gcp/leela-zero-server#15]
See:
```
2018-02-14 19:39 | ffd870d3 VS 5ac1d898 | 0 : 0 | 0 / 400 |  
2018-02-14 17:32 | b8adb7da VS 5ac1d898 | 217 : 174 (55.50%) | 391 / 400 |  
```

**Comment by WhiteHalmos** [comment]
So close!
```
2018-02-14 19:39 | ffd870d3 VS 5ac1d898 | 29 : 45 (39.19%) | 74 / 400 |  
-- | -- | -- | -- | --
2018-02-14 17:32 | b8adb7da VS 5ac1d898 | 219 : 180 (54.89%) | 399 / 400
```

**Comment by zediir** [comment]
Needs to get over 400 matches to pass because the code checks for > 220 instead of >= 220

**Comment by zediir** [comment]
>220 : 180 (55.00%) | 400 / 400 | PASS

Says it's a pass but it's not actually a pass in the database

**Comment by WhiteHalmos** [comment]
Sad. :(

**Comment by zediir** [comment]
Also there is a bug that means that even if one more victory comes in, it does not promote because for old matches the result for the net for testing promotion is fetched from the database at the same time as the incoming result is saved. This means it sometimes tests for 220 : 180 instead of 221 : 180.

**Comment by roy7** [comment]
The 55% thing I just fixed here https://github.com/gcp/leela-zero-server/commit/3b72a32b98dd6cbda6d7717b60913b08eea66cb0 oops.

I didn't realize there was a race condition on my final check, doh. Should I made check for >= 399 games? We can't be >= 55% at 399 but drop under 55% on the 400th game. 219:180 is under and 220:179 over. So if 399 is a pass, 400 must be a pass?

To fix it for real I could copy/paste this code into the two callbacks from the earlier database call, that's just a bit ugly. ;)

**Comment by WhiteHalmos** [comment]
As long as it's consistent, I think either way is fine. Thanks for putting out a fix so quickly!

If the website says PASS and then the network doesn't get promoted and fails I think that's the most discouraging. :)

**Comment by zediir** [comment]
I'd make it into a function like

```javascript
function promoteIfPass(req, new_best_network_flag, best_network_hash, cb) {
    if (!new_best_network_flag && req.body.loserhash == best_network_hash) {
        db.collection("matches").findOne({ network1: req.body.winnerhash, network2: best_network_hash, options_hash: req.body.options_hash})
        .then( (match) => { 
            if (match && ( (SPRT(match.network1_wins, match.network1_losses) === true) || (match.game_count >= 400 && match.network1_wins / match.game_count > 0.55) ) ) {
                fs.copyFileSync(__dirname + '/network/' + req.body.winnerhash + '.gz', __dirname + '/network/best-network.gz');
                console.log("New best network copied from (normal check): " + __dirname + '/network/' + req.body.winnerhash + '.gz');
            }
            cb();
        }).catch( err => {
            console.log("ERROR: " + req.body.winnerhash + " " + best_network_hash + " " + req.body.options_hash);
            console.log("ERROR: Couldn't check for new best network: " + err);
            cb(err);
        });
    }
}
```

Then call it where needed

```javascript
promoteIfPass(req, new_best_network_flag, best_network_hash, function(err) {
  // continue or return
});
```

**Comment by roy7** [comment]
Ahh yes, functions...

**Comment by Dorus** [comment]
This might be fixed now.

**Comment by roy7** [comment]
Is it? We didn't actually change this part of the code, did we? Maybe I missed it...

**Comment by Dorus** [comment]
But we moved `match.game_count++;` right? If that didn't fix it I'll look at it later. First we need to spot this error live again.

**Comment by roy7** [comment]
@Dorus The problem is around line 751. After the database insert area is done, we do another check for promotion. A race condition exists that this check could be before the prior database insert/update has completed. So we don't see the new info in the db yet, no promotion happens, then the insert/update completes. And the network should be promoted but no new check is done, until the next time a match arrives. @zediir has the right idea to make the promotion a function and call it in the 2 places needed (the database insert callbacks, so we know that query is complete before I check.

**Comment by zediir** [comment]
It might happen always if the `zlib.unzip` is a lot slower than `db.collection("matches").findOne`

**Comment by Mardak** [comment]
Sounds like this might be fixed with #64?

**Comment by roy7** [comment]
Yes indeed. I knew I'd discussed this before, but just never fixed it. There it was, Feb 17th, just lingering around. ;)
